relay-server/src/actors/project.rs

```diff
@@ -684,18 +693,18 @@ impl Project {
     /// outdated.
     pub fn enqueue_validation(&mut self, envelope: Box<Envelope>, context: EnvelopeContext) {
         match self.get_cached_state(envelope.meta().no_cache()) {
-            Some(state) => self.flush_validation(envelope, context, state),
-            None => self.pending_validations.push_back((envelope, context)),
+            Some(state) if !state.invalid() => self.flush_validation(envelope, context, state),
+            _ => self.pending_validations.push_back((envelope, context)),
         }
     }
 
     /// Adds the project state for dynamic sampling and submits the Envelope for processing.
-    fn flush_sampling(&self, mut message: ProcessEnvelope) {
+    fn flush_sampling(&self, mut message: ProcessEnvelope, project_state: Arc<ProjectState>) {
         // Intentionally ignore all errors. Fallback sampling behavior applies in this case.
-        if let Some(state) = self.valid_state() {
+        if self.valid_state().is_some() {
```

Comment:
This check is redundant now, since `get_cached_state` already checks for expiry. We could document on `flush_sampling` that this should not be called with an expired state.